Sources/NIOSSL/SSLContext.swift

```diff
@@ -779,6 +788,96 @@ extension NIOSSLContext {
+                // the user fails to retain their created and return NIOSSLContext. In the
+                // event that they do retain it, the reference count will remain at at-least
+                // 1 and thus they can re-use the same context for multiple connections.
+                let nativeContextValue = withExtendedLifetime(userChosenContext) {
+                    return CNIOBoringSSL_SSL_set_SSL_CTX(ssl, userChosenContext.sslContext)
+                }
+
+                // CNIOBoringSSL_SSL_set_SSL_CTX should return a pointer to the very same context it was
+                // told to assign on success. On error, this will not happen, so we check for equality
+                // here and handle any such errors by signaling the connection to fail.
+                if nativeContextValue != userChosenContext.sslContext {
+                    return SSL_TLSEXT_ERR_NOACK
+                }
+
+                return 1
+            }
+
+            // Make sure we get the server name extracted.
+            guard let cServerHostname = CNIOBoringSSL_SSL_get_servername(ssl, TLSEXT_NAMETYPE_host_name) else {
+                return SSL_TLSEXT_ERR_NOACK
+            }
+
+            let serverHostname = String(cString: cServerHostname)
+
+            // Can do nothing about empty indicators.
+            guard !serverHostname.isEmpty else {
+                return SSL_TLSEXT_ERR_NOACK
+            }
+
+            // Construct extension values
+            let values = NIOSSLClientExtensionValues(serverHostname: serverHostname)
+
+            // Issue call to get a new ssl context
+            let futureSSLContext: EventLoopFuture<NIOSSLContext>
+            do {
+                // This is a safe force unwrap because the context callback is always 
+                // saved in the case where we register set_cert_cb
+                futureSSLContext = try parentSwiftContext.sslContextCallback!(values, parentSwiftContext)
+            } catch {
+                // Save the user error so we can properly propagate down the pipeline
+                parentSwiftContext.sslContextCallbackResult = .failure(error)
+                return SSL_TLSEXT_ERR_NOACK
+            }
+
+            // Resume the handshake when the future completes
+            futureSSLContext.whenComplete { result in
+                // Save the result to the parent context to be inspected after resuming
+                parentSwiftContext.sslContextCallbackResult = result
+
+                // Schedule the handshake to resume. This must be scheduled on next tick to avoid resuming
+                // before suspending below. Once the handshake is resumed the entire SSL_CTX_set_cert_cb
+                // callback will be executed again, and this time the ssl context result will be saved to
+                // the parent context allowing us to set the new context.
+                futureSSLContext.eventLoop.execute {
+                    CNIOBoringSSL_SSL_do_handshake(ssl)
```

Comment:
I think this call here is a problem. This properly resumes and I see the callback running a 2nd time but nothing on the Swift side knows about it. I think this line need to somehow signal to `SSLConnection` to `doHandshake()`